<a href="https://colab.research.google.com/github/19valentynca/credit-risk-classification/blob/main/credit_risk_classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [ ]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_data = pd.read_csv("Resources/lending_data.csv")

# Review the DataFrame
lending_data.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [ ]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = lending_data['loan_status']

# Separate the X variable, the features
x = lending_data.drop(columns=['loan_status'])

In [ ]:
# Review the y variable Series
y[:5]

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [ ]:
# Review the X variable DataFrame
x[:5]

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [ ]:
# Check the balance of our target values
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [ ]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
x_train, x_test, y_train, y_test = train_test_split(x, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [ ]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
lr_model = logistic_regression_model.fit(x_train, y_train)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [ ]:
# Make a prediction using the testing data
testing_predictions = logistic_regression_model.predict(x_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [ ]:
# Print the balanced_accuracy score of the model
testing_accuracy_score = balanced_accuracy_score(y_test, testing_predictions)
print(testing_accuracy_score)

0.9520479254722232


In [ ]:
# Generate a confusion matrix for the model
testing_matrix = confusion_matrix(y_test, testing_predictions)
print(testing_matrix)

[[18663   102]
 [   56   563]]


In [ ]:
# Print the classification report for the model
testing_report = classification_report(y_test, testing_predictions)

# Print the training classification report
print(testing_report)

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model predicts our healthy loans fantastically with a precision of 100% (true positives/ (true positives + false positives)) and has a recall of 99% (true positives/ (true positives + false negatives)) meaning that the model essentially predicts heathy loan labels perfectly. Our high-risk loans paint a different picture, as we obtain a precision of 85% and a recall of 91%; still a solid prediction accuracy, but nowhere near perfect.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points.

In [ ]:
pip install -q imblearn

In [ ]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
random_oversampler_model = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
x_next, y_next = random_oversampler_model.fit_resample(x_train, y_train)
# using 'imbalanced-learn.org' found the functions and documentation for RandomOverSampler

In [ ]:
# Count the distinct values of the resampled labels data
y_next.value_counts()

0    56271
1    56271
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [ ]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
classifier = LogisticRegression(solver='lbfgs', random_state=1)

# Fit the model using the resampled training data
classifier.fit(x_next, y_next)

# Make a prediction using the testing data
predictions = classifier.predict(x_test)
pd.DataFrame({"Prediction": predictions, "Actual": y_test}).tail(10)

Prediction  Actual
73999           0       0
47267           0       0
35950           0       0
42373           0       0
38631           0       0
45639           0       0
11301           0       0
51614           0       0
4598            0       0
2793            0       0

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [ ]:
# Print the balanced_accuracy score of the model
second_accuracy_score = balanced_accuracy_score(y_test, predictions)
print(second_accuracy_score)

0.9936781215845847


In [ ]:
# Generate a confusion matrix for the model
second_matrix = confusion_matrix(y_test, predictions)
print(second_matrix)

[[18649   116]
 [    4   615]]


In [ ]:
# Print the classification report for the model
second_report = classification_report(y_test, predictions)

# Print the training classification report
print(second_report)

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.99      0.91       619

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The new logistic regression model predicts our healthy loans again fantastically with a precision of 100% (true positives/ (true positives + false positives)) and has a recall of 99% (true positives/ (true positives + false negatives)) meaning that the model essentially predicts heathy loan labels perfectly. Our high-risk loans are much different than in our first regression model. We have a lower precision at 84% (was 85% in our first regression model) but a recall of 99% (was 91% in our first regression model), meaning that we obtain a slightly higher amount of false positives in this model but a much much lower amount of false negatives than in the first model. Hence, if we were comparing the two logisitc regression models that we have carried out in this homework, I would much rather use our RandomOverSampler regression model.